From 89f8f392c3d7d4db2b0f7109e91e9133d1cafb46 Mon Sep 17 00:00:00 2001
From: Harish Kasiviswanathan <Harish.Kasiviswanathan@amd.com>
Date: Wed, 20 Dec 2017 12:21:55 -0500
Subject: [PATCH 4639/5855] drm/amdgpu: cma: fix copy within same BO

CMA copy would fail if the source and destination VAs belong to same BO.
Fix this.

BUG: SWDEV-138468

Change-Id: I7d785ad7a5c5efdf772b38486ecd423f62046a90
Signed-off-by: Harish Kasiviswanathan <Harish.Kasiviswanathan@amd.com>
---
 drivers/gpu/drm/amd/amdgpu/amdgpu_amdkfd_gpuvm.c | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/drivers/gpu/drm/amd/amdgpu/amdgpu_amdkfd_gpuvm.c b/drivers/gpu/drm/amd/amdgpu/amdgpu_amdkfd_gpuvm.c
index eb9ab84..8351558 100644
--- a/drivers/gpu/drm/amd/amdgpu/amdgpu_amdkfd_gpuvm.c
+++ b/drivers/gpu/drm/amd/amdgpu/amdgpu_amdkfd_gpuvm.c
@@ -2381,7 +2381,7 @@ int amdgpu_amdkfd_copy_mem_to_mem(struct kgd_dev *kgd, struct kgd_mem *src_mem,
 	struct amdgpu_device *adev = NULL;
 	struct amdgpu_copy_mem src, dst;
 	struct ww_acquire_ctx ticket;
-	struct list_head list;
+	struct list_head list, duplicates;
 	struct ttm_validate_buffer resv_list[2];
 	struct fence *fence = NULL;
 	int i, r;
@@ -2394,6 +2394,7 @@ int amdgpu_amdkfd_copy_mem_to_mem(struct kgd_dev *kgd, struct kgd_mem *src_mem,
 
 	adev = get_amdgpu_device(kgd);
 	INIT_LIST_HEAD(&list);
+	INIT_LIST_HEAD(&duplicates);
 
 	src.bo = &src_mem->bo->tbo;
 	dst.bo = &dst_mem->bo->tbo;
@@ -2410,7 +2411,7 @@ int amdgpu_amdkfd_copy_mem_to_mem(struct kgd_dev *kgd, struct kgd_mem *src_mem,
 		list_add_tail(&resv_list[i].head, &list);
 	}
 
-	r = ttm_eu_reserve_buffers(&ticket, &list, false, NULL);
+	r = ttm_eu_reserve_buffers(&ticket, &list, false, &duplicates);
 	if (r) {
 		pr_err("Copy buffer failed. Unable to reserve bo (%d)\n", r);
 		return r;
-- 
2.7.4

